src/python/pants/core/goals/clean.py

```diff
@@ -0,0 +1,158 @@
+            type=bool,
+            default=False,  # @TODO: True?
+            help=(
+                "Clear all caches. Can be combined with the advanced options for further control."
+            ),
+        )
+
+        # @TODO: Is there a way to collect these programmatically?
+        for flag_slug, help in _NAMED_CACHE_FLAGS_INFO:
+            register(
+                f"--keep-{flag_slug}",
+                advanced=True,
+                type=bool,
+                default=True,
+                help=help,
+            )
+
+
+class Clean(Goal):
+    subsystem_cls = CleanSubsystem
+
+
+@goal_rule
+async def clean(
+    clean_subsystem: CleanSubsystem,
+    console: Console,
+    real_opts: _Options,
+    dist_dir: DistDir,
+) -> Clean:
+    def _maybe_clean(
+        keep_flag_slug: str,
+        keep_flag_value: bool,
+        path: Union[str, Path],
+        glob: str = "*",
+    ):
+        path = Path(path)
+        paths = list(path.glob(glob))
+        should_clean = (clean_subsystem.options.all and not keep_flag_value) or (
+            clean_subsystem.options.all and not not keep_flag_value
+        )
+
+        if clean_subsystem.options.dry_run:
+            # @TODO: handle empty (0 bytes or 0 files) specially?
+            total_size, num_files = _dir_info(*paths)
+            maybe_negative = "" if should_clean else console.red("not ")
+
+            console.print_stdout(f"Would {maybe_negative}clean {console.green(path / glob)}:")
+            # @TODO: Print the relevant flag that gave us this directory?
+            console.print_stdout(f"  Controlled by {console.blue(f'--[no-]keep-{keep_flag_slug}')}")
+            console.print_stdout(
+                f"  {len(paths)} dirs, spanning {humanize.intword(num_files)} files totalling {humanize.naturalsize(total_size)}"
+            )
+        elif should_clean:
+            # @TODO: Should we multi-process this?
+            console.print_stdout(f"Cleaning {console.green(path / glob)}")
+            for path in paths:
+                rm_rf(str(path))
+
+    global_options = real_opts.options.for_global_scope()
+    bootstrap_options = real_opts.options.bootstrap_option_values()
+    pants_cache_dir = get_pants_cachedir()
+
+    for flag_slug, _ in _NAMED_CACHE_FLAGS_INFO:
+        _maybe_clean(
+            flag_slug,
+            getattr(clean_subsystem.options, f"keep_{flag_slug.replace('-', '_')}"),
+            *{
+                "process-execution-temp-dirs": (Path(tempfile.gettempdir()), "pants-pe*"),
+                "lmdb-store": (LocalStoreOptions.from_options(global_options).store_dir,),
+                "run-tracker": (Path(global_options.pants_workdir) / "run-tracker",),
+                "repl-temp-dirs": (Path(global_options.pants_workdir), "repl*"),
+                "dist-dir": (dist_dir.relpath,),
+                "bootstrap-dir": (Path(pants_cache_dir) / "setup",),
+                "named-caches": (bootstrap_options.named_caches_dir,),
```

Comment:
Maybe move these as `lambdas` into the constant you have? Maybe with a `Protocol` that defines the arguments (`GlobalOptions`, etc).